core/includes/install.core.inc

```diff
@@ -1454,6 +1450,341 @@ function install_select_language_form($form, &$form_state, $files) {
   return $form;
 }
 
+/**
+ * Provide a list of available translation files.
+ *
+ * @return array
+ *   An array of language codes.
+ */
+function install_get_available_translations($files) {
+  $available_translations = array();
+  $response = backdrop_http_request(INSTALL_AVAILABLE_TRANSLATIONS_URI);
+
+  $data = '';
+  if ($response->code === '200') {
+    $data = json_decode($response->data, TRUE);
+  }
+
+  if ($data) {
+    $available_translations = array_keys($data);
+  }
+  elseif (count($files) > 1) {
+    foreach($files as $translation_file) {
+      $available_translations[] = $translation_file->langcode;
+    }
+  }
+  else {
+    return FALSE;
+  }
+
+  // English is always available.
+  $available_translations[] = 'en';
+  $standard_languages = standard_language_list();
+  return array_intersect_key($standard_languages, array_flip(array_unique($available_translations)));
+}
+
+/**
+ * Download a translation file for the selected language.
+ *
+ * @param array $install_state
+ *   An array of information about the current installation state.
+ *
+ * @return string
+ *   A themed status report, or an exception if there are requirement errors.
+ *   Upon successful download the page is reloaded and no output is returned.
+ */
+function install_download_translation(&$install_state) {
+  // Check whether all conditions are met to download. Download the translation
+  // if possible.
+  $requirements = install_check_translations($install_state['parameters']['langcode'], $install_state['server_pattern']);
+  // Render requirements if any warnings or errors returned.
+  if ($output = install_display_requirements($install_state, $requirements)) {
+    return $output;
+  }
+
+  // The download was successful, reload the page in the new language.
+  $install_state['translations'][$install_state['parameters']['langcode']] = TRUE;
+}
+
+/**
+ * Attempts to get a file using a HTTP request and to store it locally.
+ *
+ * @param string $uri
+ *   The URI of the file to grab.
+ * @param string $destination
+ *   Stream wrapper URI specifying where the file should be placed. If a
+ *   directory path is provided, the file is saved into that directory under its
+ *   original name. If the path contains a filename as well, that one will be
+ *   used instead.
+ *
+ * @return bool
+ *   TRUE on success, FALSE on failure.
+ */
+function install_retrieve_file($uri, $destination) {
+  $parsed_url = parse_url($uri);
+
+  if (is_dir(backdrop_realpath($destination))) {
+    // Prevent URIs with triple slashes when gluing parts together.
+    $path = str_replace('///', '//', "$destination/") . backdrop_basename($parsed_url['path']);
+  }
+  else {
+    $path = $destination;
+  }
+
+  $response = backdrop_http_request($uri);
```

Comment:
backdrop_http_request() allows for the timeout to be set. The default is 30 seconds - that's quite long.

How about... 10 seconds instead? (Just a quick idea).
Otherwise people have to wait quite looooong for a "can't connect, anyway". :wink:

Reply:
I agree 👍🏼 ...but IMO, 10sec is still too long. Even on slow connections, something like 5sec should be enough to get a response - anything else should indicate some problem, in which case we should be falling back to English, showing a message, and allowing people to move on using English.

If you google with something like "what is acceptable web page load time", you get responses like this:
> An ideal page load time is between 0-2 seconds, but 3 seconds is also considered to be an acceptable score. Anything above 3 seconds increases the likelihood of visitors leaving your site.

Reply:
Fixed, used 5s as default.